cmake: Add OmrOption, a util for declaring build options

rwy0717 opened the pull request:
OmrOption is a module which provides a declarative syntax for specifying
and validating ON/OFF build options. The long-term goal is to declare
and validate all OMR build options using this module.

The general syntax is:
```
  omr_add_option_group(<group>)

  omr_add_option(<option> <option-group>
    COMMENT <comment>
    DEFAULT <expression>
    REQUIRES <option>...
    REQUIRES_EXPR <expr>
    CONFLICTS <option>...
  )

  omr_validate_option_group(<group>)
```
Options are defined as BOOL cache variables.

DEFAULT is an expression, which is evaluated and coerced to an on/off
value. The default value is only evaluated on the first run, and cached
in subsequent runs. Since options are regular cmake cache variables, the
value can be manually overridden by users at any point. A default value
must always be provided.

REQUIRES allows users to pass a list of required flags or variables,
which must be ON if the option is ON.

Similarly, CONFLICTS is a list of options which must be OFF, if the
option is ON.

REQUIRES_EXPR is an expression which must be TRUE, if the option is
ON. This can be used for expressing more complex option requirements.
Each option can have only one associated REQUIRES_EXPR.

Options are registered into a group, which must be explicitly validated
with a call to omr_validate_option_group. Option groups allow consuming
projects to create and validate their own option sets. By making
validation explicit, we allow the user (us OMR developers) to delay
validation until after all configuration variables have been set.
Options are validated on every run of cmake, which ensures that the user
never puts OMR into an illegal configuration.

The phrasing "option" was chosen over "flag" to follow existing cmake
conventions, where add_option / add_dependent_option create ON/OFF
bool cache variables.

There is a similar module built in to cmake called DependentOption.
This module was not used because:
1. A dependent option is only show to the user if it's dependencies are
   satisfied. OmrOption makes the choice to always present ALL flags in
   the cache.
2. A dependent option will revert to it's default value if it's
   requirements aren't satisfied, overriding any user provided value.
   OmrOption never tries to override the user, and relies on option
   validation instead.
3. OmrOption provides an easier to read, more obvious syntax.
4. The default value in OmrOption is an expression, rather than a simple
   on/off constant. This gives us a succinct way to implement implicitly
   enabled/disabled flags, while still being user overridable.
   Dependent options have a dependency expression, but the default value
   is a constant, which cannot be overridden by the user.

Signed-off-by: Robert Young <rwy0717@gmail.com>

rwy0717 commented:
/cc @Leonardo2718 @youngar @dnakamura

rwy0717 commented:
(updated some comments)

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
I think some of the content from your PR description would be good to also have here (especially the comparison with `DependentOption`) so that we don't need to look back at the PR to find that of information in the future.

Leonardo2718 replied:
Also, for completeness, can you add an example of actually reading one of the defined options?

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Can you add some guidelines for how this function should be used? It's not clear to me whether this is only intended for use in the implementation of this utility or for general, "public" use. If the latter, it might also be useful to add a short example.

rwy0717 replied:
For the most part, I wrote these functions to be publicly callable... although there are a few where it doesn't always make sense, like `omr_check_option_requirements `. There aren't any conventions in cmake for private functions, so I suppose the right thing to do is document the complete public API in the module's documentation comment. I can move the signature comments for public functions up into the module docs, if that works?

Leonardo2718 replied:
Sure, I think that works. Maybe also add an example usages, where it makes sense?

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
Same comment as above.

rwy0717 commented:
I was thinking about adding some tests, but I'm not sure how to actually implement them. I can look at using `add_test` to run cmake scripts, which can use omr_assert to check that the module is functioning correctly. However, it would be difficult to test invalid configurations--I don't know how to check that a script aborted in the "right" way.

Leonardo2718 commented:
> However, it would be difficult to test invalid configurations--I don't know how to check that a script aborted in the "right" way.

Hm... yeah that does seem a bit tricky.

rwy0717 commented:
changes:
1. added lots of documentation to the header
2. added a leading underscore to internal function names
3. allow `omr_add_option_group` to declare multiple groups at once
4. slightly modified error output
5. fixed bug in `omr_validate_option`: correctly check `if(${option})`, not `if(option)`.

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```diff
@@ -0,0 +1,372 @@
+  omr_option_conflicts(<option> [<conflict>...])
+
+Add conflicts to ``<option>``. If the option is ``ON``, each variable named
+in the conflicts must _not_ be true.
```
Would it be clearer to just say "must be false"?

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```diff
@@ -0,0 +1,372 @@
+
+Create a new option. The new option is a ``BOOL`` cache variable.
+The option will be registered in ``<group>``. When ``<group>`` is
+validated, this options requirements will be tested.
```
Should `options` be `option's`?

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```diff
@@ -0,0 +1,372 @@
+
+Register a set of options into an options group. Options must have already
+been declared. It's possible (but not recommended) to register an option in
+to multiple groups.
```
"in to" &rarr; "into"

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```suggestion
1. A dependent option is only shown to the user if it's dependencies are
```

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```suggestion
# _omr_get_options_group(<output> <group>)
```

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```suggestion
# _omr_check_option_requirements(<option>)
```

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```suggestion
# _omr_check_option_conflicts(<option>)
```

Leonardo2718 commented on cmake/modules/OmrOption.cmake:
```suggestion
# _omr_check_option_requirement_expr(<option>)
```
